### DISCUSSION TITLE
Allow running programs that have compiler errors

### DISCUSSION THREAD

**Discussion by qrzychu** [dotnet/roslyn#62535] [General]
I've been following Richard Feldman's Roc-Lang journey, and in one of his talks he said this: https://youtu.be/6qzWm_eoUXM?t=2939

The Roc compiler is able to emit a working program, even with compiler errors. It's just that if the runtime hits a point where the didn't compile (like a type mismatch, missing variable, incomplete pattern, incompatible function parameter list etc), the program crashes.

He then goes to explain why this is usefull - for example unit testing. If the parser is binary (either the code passes or it does not), refactoring and testing of the refactored code can be hindered by the need to fix the whole codebase, not just to code related to a single unit test.

I can image the productivity boost of something like that combined with hot reload for GUI apps for example. That's why people like using for Python or JavaScript, because if they want, they can just focus on one single path, test it, then figure out how to do the rest, while we with C# need to figure it all out before we can even check if we parsed inputs correctly (I know, with seperation of concerns you can kind of get past that, but if you you change the return type of `ParseInputs` you have to fix all usages (could be hundreds) and generally fix all compiler errros in the whole project before you can run any unit test against it)

The question is, is running incorrect programs even technically possible with C# and CLR? Maybe there could be some kind of "interpreter mode" that you could toggle on for unit testing? Of course in `Release` mode, no program.

Or maybe just straight up replace every error with something like `Assert.NotCompiled`?

Or maybe just don't compile code that can't be reached because of the errors?

I have no idea how it could work in dotnet context, just wanted to let you know that someone somewhere is working on that :) I am also aware, that it some cases the resulting program would be extremely short and useless, but sometime it would be nice to shed the static type checker just for a bit.

**Comment by CyrusNajmabadi** [comment]
Moving to roslyn.  Thsi is not a concern of the language.  Indeed, the language does not say that an invalid program should not be runnable.

**Reply by HaloFour** [reply]
I'm pretty sure there is an existing issue/discussion on this, but I can't seem to find it.

**Reply by PathogenDavid** [reply]
I managed to dig it up thanks to Rikki's mentioning of "errors as warnings": https://github.com/dotnet/roslyn/issues/305

The original discussion was on Codeplex, here's a quick-and-dirty copy of that discussion extracted from the archives: [Codeplex#573979](https://gist.github.com/PathogenDavid/40f57ffd28fc3df3f76d751b24838fb0)

**Reply by qrzychu** [reply]
So, after going through those discussion, I end up at reference assemblies ([docs](https://docs.microsoft.com/en-us/dotnet/standard/assembly/reference-assemblies)).

However, that's not exactly what I meant. Reference assemblies are "empty shells" that can describe the shape of a real assembly.

So, if I have a class:

```csharp
public class Class1
{
   public string Method1() => "method 1";
  public string Method2() => 5; // this is a compilation error
}
```

I expected being able to actually run `class1Instance.Method1()`, not just being able to reference this asembly in another csproj and compile the other project. I want to be able to run a unit test agianst `Method1` that has correct implementation, while `Method2` has incorrect code.

This should run and pass:

```csharp
[Fact]
public void Test()
{
   new Class1().Method1().Should().Be("method 1");  // this should be ok, since all needed code compiles
}

[Fact]
public void Test2()
{
   new Class1().Method2().Should().Be("method 2"); // this should fail on runtime with something like `MethodNotImplemented`
}
```

Reference assemblies don't allow that, as far as I understand that.

**Comment by tmat** [comment]
FWIW - Hot Reload and EnC allow applying edits even if there are compilation errors elsewhere in the code as long as the edited code is clean.

**Comment by RikkiGibson** [comment]
IIRC, the name of this feature is "errors as warnings" 😉

**Reply by qrzychu** [reply]
that's the perfect name for something like this :D
